File: modules/moonshine-voice/ios/MoonshineVoiceModule.swift
import ExpoModulesCore
import MoonshineVoice

/// Expo native module bridging Moonshine Voice to React Native.
/// Provides streaming mic transcription, file transcription, and intent recognition.
public class MoonshineVoiceModule: Module {
  private var transcriber: Transcriber?
  private var micTranscriber: MicTranscriber?
  private var intentRecognizer: IntentRecognizer?

  /// Accumulated text from the current streaming session.
  private var streamingText: String = ""
  private var isStreaming: Bool = false

  public func definition() -> ModuleDefinition {
    Name("MoonshineVoice")

    // MARK: - Events

    Events(
      "onPartialTranscript",
      "onLineCompleted",
      "onIntentRecognized"
    )

    // MARK: - Sync functions

    Function("isAvailable") {
      return true
    }

    Function("isModelLoaded") {
      return self.transcriber != nil
    }

    // MARK: - Async functions

    AsyncFunction("loadModel") { (modelPath: String, modelArch: Int) -> Bool in
      return try await self.doLoadModel(modelPath: modelPath, modelArch: modelArch)
    }

    AsyncFunction("loadBundledModel") { (modelArch: Int) -> Bool in
      guard let bundlePath = Bundle.main.resourcePath else {
        throw MoonshineError.modelNotFound
      }
      let modelDir = (bundlePath as NSString).appendingPathComponent("moonshine-models")
      return try await self.doLoadModel(modelPath: modelDir, modelArch: modelArch)
    }

    AsyncFunction("unloadModel") {
      self.stopMicIfNeeded()
      self.intentRecognizer = nil
      self.transcriber = nil
    }

    AsyncFunction("startStreaming") {
      guard let transcriber = self.transcriber else {
        throw MoonshineError.modelNotLoaded
      }
      guard !self.isStreaming else {
        throw MoonshineError.alreadyStreaming
      }

      self.streamingText = ""
      self.isStreaming = true

      let mic = MicTranscriber(transcriber: transcriber)
      mic.addListener(self.makeTranscriptListener())
      if let ir = self.intentRecognizer {
        mic.addListener(ir)
      }
      try mic.start()
      self.micTranscriber = mic
    }

    AsyncFunction("stopStreaming") { () -> String in
      self.stopMicIfNeeded()
      let result = self.streamingText
      self.streamingText = ""
      self.isStreaming = false
      return result
    }

    AsyncFunction("transcribeFile") { (filePath: String) -> [String: Any] in
      guard let transcriber = self.transcriber else {
        throw MoonshineError.modelNotLoaded
      }

      let url = URL(fileURLWithPath: filePath)
      let result = try transcriber.transcribeFile(url)

      return [
        "text": result.text,
        "duration": result.duration,
        "segments": result.lines.map { line in
          [
            "text": line.text,
            "startTime": line.startTime,
            "duration": line.duration,
          ] as [String: Any]
        },
      ]
    }

    AsyncFunction("registerIntent") { (triggerPhrase: String, intentName: String) in
      guard let ir = self.intentRecognizer else {
        throw MoonshineError.intentRecognizerNotLoaded
      }
      ir.registerIntent(triggerPhrase: triggerPhrase, name: intentName) {
        [weak self] phrase, utterance, similarity in
        self?.sendEvent("onIntentRecognized", [
          "intent": phrase,
          "utterance": utterance,
          "similarity": similarity,
        ])
      }
    }

    AsyncFunction("unregisterIntent") { (intentName: String) in
      self.intentRecognizer?.unregisterIntent(name: intentName)
    }

    AsyncFunction("clearIntents") {
      self.intentRecognizer?.clearIntents()
    }
  }

  // MARK: - Private helpers

  private func doLoadModel(modelPath: String, modelArch: Int) throws -> Bool {
    let arch = self.mapArch(modelArch)
    let t = try Transcriber(modelPath: modelPath, modelArch: arch)
    self.transcriber = t

    // Try to load intent recognizer (Gemma embeddings) — optional, non-fatal if missing
    if let embeddingPath = self.findEmbeddingModel(near: modelPath) {
      self.intentRecognizer = try? IntentRecognizer(
        modelPath: embeddingPath,
        modelArch: .gemma300M,
        threshold: 0.75
      )
    }

    return true
  }

  private func mapArch(_ raw: Int) -> ModelArch {
    switch raw {
    case 0: return .tiny
    case 1: return .base
    case 2: return .tinyStreaming
    case 3: return .baseStreaming
    case 4: return .smallStreaming
    case 5: return .mediumStreaming
    default: return .tinyStreaming
    }
  }

  private func findEmbeddingModel(near modelPath: String) -> String? {
    let parent = (modelPath as NSString).deletingLastPathComponent
    let embeddingDir = (parent as NSString).appendingPathComponent("gemma-embedding")
    if FileManager.default.fileExists(atPath: embeddingDir) {
      return embeddingDir
    }
    return nil
  }

  private func stopMicIfNeeded() {
    micTranscriber?.stop()
    micTranscriber = nil
  }

  private func makeTranscriptListener() -> TranscriptEventListener {
    return ModuleTranscriptListener(module: self)
  }

  /// Bridge class that forwards Moonshine transcript events to JS.
  fileprivate class ModuleTranscriptListener: TranscriptEventListener {
    weak var module: MoonshineVoiceModule?

    init(module: MoonshineVoiceModule) {
      self.module = module
    }

    func onLineTextChanged(_ event: TranscriptLineEvent) {
      module?.streamingText = event.line.text
      module?.sendEvent("onPartialTranscript", [
        "text": event.line.text,
        "isComplete": event.line.isComplete,
        "lineId": event.line.lineId,
        "startTime": event.line.startTime,
      ])
    }

    func onLineCompleted(_ event: TranscriptLineEvent) {
      module?.sendEvent("onLineCompleted", [
        "text": event.line.text,
        "startTime": event.line.startTime,
        "duration": event.line.duration,
        "lineId": event.line.lineId,
      ])
    }
  }
}

// MARK: - Errors

enum MoonshineError: Error, LocalizedError {
  case modelNotFound
  case modelNotLoaded
  case alreadyStreaming
  case intentRecognizerNotLoaded

  var errorDescription: String? {
    switch self {
    case .modelNotFound:
      return "Moonshine model files not found in app bundle."
    case .modelNotLoaded:
      return "No Moonshine model is loaded. Call loadModel() first."
    case .alreadyStreaming:
      return "Streaming is already active. Call stopStreaming() first."
    case .intentRecognizerNotLoaded:
      return "Intent recognizer not available. Embedding model may be missing."
    }
  }
}

File: modules/moonshine-voice/android/src/main/java/com/routemasterpro/moonshine/MoonshineVoiceModule.kt
package com.routemasterpro.moonshine

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import ai.moonshine.voice.MicTranscriber
import ai.moonshine.voice.Transcriber
import ai.moonshine.voice.TranscriptEventListener
import ai.moonshine.voice.TranscriptLineEvent
import ai.moonshine.voice.IntentRecognizer
import ai.moonshine.voice.ModelArch
import ai.moonshine.voice.EmbeddingModelArch
import java.io.File

class MoonshineVoiceModule : Module() {
  private var transcriber: Transcriber? = null
  private var micTranscriber: MicTranscriber? = null
  private var intentRecognizer: IntentRecognizer? = null
  private var streamingText: String = ""
  private var isStreaming: Boolean = false

  override fun definition() = ModuleDefinition {
    Name("MoonshineVoice")

    Events("onPartialTranscript", "onLineCompleted", "onIntentRecognized")

    Function("isAvailable") { true }

    Function("isModelLoaded") { transcriber != null }

    AsyncFunction("loadModel") { modelPath: String, modelArch: Int ->
      doLoadModel(modelPath, modelArch)
    }

    AsyncFunction("loadBundledModel") { modelArch: Int ->
      val context = appContext.reactContext ?: throw CodedException("NO_CONTEXT", "App context unavailable", null)
      // Load from assets: models are expected in assets/moonshine-models/
      val t = Transcriber.loadFromAssets(context, "moonshine-models", mapArch(modelArch))
      transcriber = t

      // Try loading intent recognizer from assets (optional)
      try {
        intentRecognizer = IntentRecognizer.loadFromAssets(
          context, "gemma-embedding", EmbeddingModelArch.GEMMA_300M, 0.75f
        )
      } catch (_: Exception) { /* Embedding model not bundled — skip */ }

      true
    }

    AsyncFunction("unloadModel") {
      stopMicIfNeeded()
      intentRecognizer = null
      transcriber = null
    }

    AsyncFunction("startStreaming") {
      val t = transcriber ?: throw CodedException("NOT_LOADED", "No model loaded. Call loadModel() first.", null)
      if (isStreaming) throw CodedException("ALREADY_STREAMING", "Streaming already active.", null)

      streamingText = ""
      isStreaming = true

      val context = appContext.reactContext ?: throw CodedException("NO_CONTEXT", "App context unavailable", null)
      val mic = MicTranscriber(t)
      mic.addListener(createTranscriptListener())
      intentRecognizer?.let { mic.addListener(it) }
      mic.grantPermissions() // Signal that permissions are handled by Expo
      mic.start()
      micTranscriber = mic
    }

    AsyncFunction("stopStreaming") {
      stopMicIfNeeded()
      val result = streamingText
      streamingText = ""
      isStreaming = false
      result
    }

    AsyncFunction("transcribeFile") { filePath: String ->
      val t = transcriber ?: throw CodedException("NOT_LOADED", "No model loaded.", null)
      val result = t.transcribeWithoutStreaming(File(filePath))
      mapOf(
        "text" to (result?.text ?: ""),
        "duration" to (result?.duration ?: 0.0),
        "segments" to (result?.lines?.map { line ->
          mapOf(
            "text" to line.text,
            "startTime" to line.startTime,
            "duration" to line.duration,
          )
        } ?: emptyList<Map<String, Any>>()),
      )
    }

    AsyncFunction("registerIntent") { triggerPhrase: String, intentName: String ->
      val ir = intentRecognizer
        ?: throw CodedException("NO_INTENT", "Intent recognizer not available.", null)
      ir.registerIntent(triggerPhrase, intentName) { phrase, utterance, similarity ->
        sendEvent("onIntentRecognized", mapOf(
          "intent" to phrase,
          "utterance" to utterance,
          "similarity" to similarity,
        ))
      }
    }

    AsyncFunction("unregisterIntent") { intentName: String ->
      intentRecognizer?.unregisterIntent(intentName)
    }

    AsyncFunction("clearIntents") {
      intentRecognizer?.clearIntents()
    }
  }

  private fun doLoadModel(modelPath: String, modelArch: Int): Boolean {
    val t = Transcriber.loadFromFiles(modelPath, mapArch(modelArch))
    transcriber = t

    // Try to load intent recognizer from nearby embedding model dir
    val embeddingDir = File(File(modelPath).parent, "gemma-embedding")
    if (embeddingDir.exists()) {
      try {
        intentRecognizer = IntentRecognizer(
          embeddingDir.absolutePath, EmbeddingModelArch.GEMMA_300M, 0.75f
        )
      } catch (_: Exception) { /* Non-fatal */ }
    }

    return true
  }

  private fun mapArch(raw: Int): ModelArch = when (raw) {
    0 -> ModelArch.TINY
    1 -> ModelArch.BASE
    2 -> ModelArch.TINY_STREAMING
    3 -> ModelArch.BASE_STREAMING
    4 -> ModelArch.SMALL_STREAMING
    5 -> ModelArch.MEDIUM_STREAMING
    else -> ModelArch.TINY_STREAMING
  }

  private fun stopMicIfNeeded() {
    micTranscriber?.stop()
    micTranscriber = null
  }

  private fun createTranscriptListener(): TranscriptEventListener {
    return object : TranscriptEventListener {
      override fun onLineTextChanged(event: TranscriptLineEvent) {
        streamingText = event.line.text
        sendEvent("onPartialTranscript", mapOf(
          "text" to event.line.text,
          "isComplete" to event.line.isComplete,
          "lineId" to event.line.lineId,
          "startTime" to event.line.startTime,
        ))
      }

      override fun onLineCompleted(event: TranscriptLineEvent) {
        sendEvent("onLineCompleted", mapOf(
          "text" to event.line.text,
          "startTime" to event.line.startTime,
          "duration" to event.line.duration,
          "lineId" to event.line.lineId,
        ))
      }
    }
  }
}
